src/mbgl/style/expression/compound_expression.cpp

```diff
@@ -44,35 +78,31 @@ struct Signature;
 // Simple evaluate function (const T0&, const T1&, ...) -> Result<U>
 template <class R, class... Params>
 struct Signature<R (Params...)> : SignatureBase {
-    using Args = std::array<std::unique_ptr<Expression>, sizeof...(Params)>;
-
     Signature(R (*evaluate_)(Params...), std::string name_) :
         SignatureBase(
             valueTypeToExpressionType<std::decay_t<typename R::Value>>(),
             std::vector<type::Type> {valueTypeToExpressionType<std::decay_t<Params>>()...},
             std::move(name_)
         ),
         evaluate(evaluate_)    {}
-    
-    EvaluationResult apply(const EvaluationContext& evaluationParameters, const Args& args) const {
-        return applyImpl(evaluationParameters, args, std::index_sequence_for<Params...>{});
-    }
 
-    std::unique_ptr<Expression> makeExpression(std::vector<std::unique_ptr<Expression>> args) const override {
-        typename Signature::Args argsArray;
-        std::copy_n(std::make_move_iterator(args.begin()), sizeof...(Params), argsArray.begin());
-        return std::make_unique<CompoundExpression<Signature>>(name, *this, std::move(argsArray));
+    EvaluationResult apply(const EvaluationContext& evaluationParameters, const Args& args) const override {
+        return applyImpl(evaluationParameters, args, std::index_sequence_for<Params...>{});
     }
 
     R (*evaluate)(Params...);
+
 private:
     template <std::size_t ...I>
     EvaluationResult applyImpl(const EvaluationContext& evaluationParameters, const Args& args, std::index_sequence<I...>) const {
-        const std::array<EvaluationResult, sizeof...(I)> evaluated = {{std::get<I>(args)->evaluate(evaluationParameters)...}};
-        for (const auto& arg : evaluated) {
-            if(!arg) return arg.error();
```

Comment:
What do the benchmarks look like after this change?  IIRC, the reason for it (which I should have left a comment about) was to avoid the overhead of using `vector` for the very common case of expressions with a known number of arguments (especially: `get`, `zoom`)

Reply:
Hard to tell.

Before:

```
-------------------------------------------------------------------------------------
Benchmark                                              Time           CPU Iterations
-------------------------------------------------------------------------------------
API_queryRenderedFeaturesAll                   300407698 ns  296336000 ns          2
API_queryRenderedFeaturesLayerFromLowDensity      508258 ns     502216 ns       1220
API_queryRenderedFeaturesLayerFromHighDensity    7374136 ns    7312417 ns         96
API_renderStill_reuse_map                       17518497 ns   13231553 ns         47
API_renderStill_reuse_map_switch_styles        193390479 ns  167888333 ns          3
API_renderStill_recreate_map                   262035790 ns  205813000 ns          4
Parse_CameraFunction/1                              5305 ns       5307 ns     126975 1
Parse_CameraFunction/2                              6095 ns       6095 ns     114651 2
Parse_CameraFunction/4                              6972 ns       6978 ns      95615 4
Parse_CameraFunction/6                              8194 ns       8186 ns      85157 6
Parse_CameraFunction/8                              9157 ns       9155 ns      76182 8
Parse_CameraFunction/10                            10425 ns      10399 ns      66310 10
Parse_CameraFunction/12                            11470 ns      11452 ns      60637 12
Evaluate_CameraFunction/1                             48 ns         48 ns   14742087 1
Evaluate_CameraFunction/2                             88 ns         88 ns    7917209 2
Evaluate_CameraFunction/4                            109 ns        109 ns    6417014 4
Evaluate_CameraFunction/6                            121 ns        120 ns    5882304 6
Evaluate_CameraFunction/8                            122 ns        121 ns    5737564 8
Evaluate_CameraFunction/10                           128 ns        127 ns    5389632 10
Evaluate_CameraFunction/12                           130 ns        129 ns    5400651 12
Parse_CompositeFunction/1                           8293 ns       8294 ns      83866 1
Parse_CompositeFunction/2                          13110 ns      13093 ns      53193 2
Parse_CompositeFunction/4                          28377 ns      28268 ns      24930 4
Parse_CompositeFunction/6                          50902 ns      50613 ns      13574 6
Parse_CompositeFunction/8                          81914 ns      81346 ns       8485 8
Parse_CompositeFunction/10                        117934 ns     117284 ns       5951 10
Parse_CompositeFunction/12                        160801 ns     159857 ns       4351 12
Evaluate_CompositeFunction/1                         397 ns        392 ns    1798409 1
Evaluate_CompositeFunction/2                         547 ns        544 ns    1316383 2
Evaluate_CompositeFunction/4                         626 ns        621 ns    1121849 4
Evaluate_CompositeFunction/6                         674 ns        668 ns    1026318 6
Evaluate_CompositeFunction/8                         690 ns        686 ns    1051509 8
Evaluate_CompositeFunction/10                        687 ns        684 ns    1017191 10
Evaluate_CompositeFunction/12                        712 ns        708 ns     997435 12
Parse_SourceFunction/1                              6680 ns       6679 ns     100838 1
Parse_SourceFunction/2                              7389 ns       7392 ns      95060 2
Parse_SourceFunction/4                              8365 ns       8364 ns      82945 4
Parse_SourceFunction/6                              9413 ns       9408 ns      74319 6
Parse_SourceFunction/8                             10514 ns      10514 ns      67157 8
Parse_SourceFunction/10                            11546 ns      11529 ns      58893 10
Parse_SourceFunction/12                            12595 ns      12585 ns      53498 12
Evaluate_SourceFunction/1                            367 ns        364 ns    1936783 1
Evaluate_SourceFunction/2                            409 ns        407 ns    1735620 2
Evaluate_SourceFunction/4                            431 ns        429 ns    1620299 4
Evaluate_SourceFunction/6                            450 ns        446 ns    1578091 6
Evaluate_SourceFunction/8                            446 ns        444 ns    1583187 8
Evaluate_SourceFunction/10                           456 ns        453 ns    1197379 10
Evaluate_SourceFunction/12                           452 ns        450 ns    1556119 12
Parse_Filter                                        4879 ns       4856 ns     142963
Parse_EvaluateFilter                                 119 ns        119 ns    5931952
TileMaskGeneration                                  1796 ns       1780 ns     354100
Parse_VectorTile                                 1913299 ns    1904227 ns        361
Util_dtoa                                           2052 ns       2039 ns     339971
Util_standardDtoa                                   2695 ns       2683 ns     259530
Util_dtoaLimits                                      309 ns        307 ns    2248093
Util_standardDtoaLimits                            35771 ns      35611 ns      19544
TileCountBounds                                       93 ns         92 ns    7496011
TileCountPolygon                                   22679 ns      22574 ns      30130
TileCoverPitchedViewport                            3302 ns       3286 ns     210440
TileCoverBounds                                     1240 ns       1234 ns     556182
TileCoverPolygon                                    8890 ns       8834 ns      78466
```

After:

```
-------------------------------------------------------------------------------------
Benchmark                                              Time           CPU Iterations
-------------------------------------------------------------------------------------
API_queryRenderedFeaturesAll                   431325626 ns  400442500 ns          2
API_queryRenderedFeaturesLayerFromLowDensity      760784 ns     711756 ns        864
API_queryRenderedFeaturesLayerFromHighDensity   11047729 ns   10367304 ns         69
API_renderStill_reuse_map                       39430648 ns   30756176 ns         17
API_renderStill_reuse_map_switch_styles        396607019 ns  291833500 ns          2
API_renderStill_recreate_map                   313133398 ns  250921667 ns          3
Parse_CameraFunction/1                              5554 ns       5535 ns     121597 1
Parse_CameraFunction/2                              6106 ns       6093 ns     110972 2
Parse_CameraFunction/4                              7199 ns       7183 ns      96657 4
Parse_CameraFunction/6                              8358 ns       8342 ns      81146 6
Parse_CameraFunction/8                              9515 ns       9490 ns      72348 8
Parse_CameraFunction/10                            10745 ns      10702 ns      65701 10
Parse_CameraFunction/12                            11829 ns      11765 ns      57713 12
Evaluate_CameraFunction/1                             49 ns         49 ns   14267138 1
Evaluate_CameraFunction/2                             92 ns         91 ns    7618301 2
Evaluate_CameraFunction/4                            115 ns        114 ns    6190581 4
Evaluate_CameraFunction/6                            124 ns        124 ns    5609604 6
Evaluate_CameraFunction/8                            127 ns        126 ns    5546487 8
Evaluate_CameraFunction/10                           132 ns        131 ns    5325989 10
Evaluate_CameraFunction/12                           134 ns        133 ns    5241012 12
Parse_CompositeFunction/1                           8136 ns       8128 ns      85387 1
Parse_CompositeFunction/2                          12540 ns      12491 ns      55030 2
Parse_CompositeFunction/4                          26914 ns      26773 ns      26185 4
Parse_CompositeFunction/6                          48899 ns      48633 ns      14392 6
Parse_CompositeFunction/8                          78678 ns      78193 ns       8964 8
Parse_CompositeFunction/10                        113316 ns     112756 ns       5979 10
Parse_CompositeFunction/12                        157886 ns     156933 ns       4421 12
Evaluate_CompositeFunction/1                         530 ns        527 ns    1282662 1
Evaluate_CompositeFunction/2                         743 ns        737 ns     972182 2
Evaluate_CompositeFunction/4                         852 ns        846 ns     836990 4
Evaluate_CompositeFunction/6                         890 ns        884 ns     771886 6
Evaluate_CompositeFunction/8                         914 ns        908 ns     773233 8
Evaluate_CompositeFunction/10                        924 ns        918 ns     763109 10
Evaluate_CompositeFunction/12                        942 ns        935 ns     748679 12
Parse_SourceFunction/1                              6571 ns       6561 ns     103510 1
Parse_SourceFunction/2                              7247 ns       7230 ns      98182 2
Parse_SourceFunction/4                              8244 ns       8209 ns      83879 4
Parse_SourceFunction/6                              9413 ns       9365 ns      74458 6
Parse_SourceFunction/8                             10534 ns      10475 ns      63881 8
Parse_SourceFunction/10                            11588 ns      11544 ns      58441 10
Parse_SourceFunction/12                            12684 ns      12606 ns      54277 12
Evaluate_SourceFunction/1                            516 ns        509 ns    1330065 1
Evaluate_SourceFunction/2                            561 ns        554 ns    1267565 2
Evaluate_SourceFunction/4                            572 ns        567 ns    1151354 4
Evaluate_SourceFunction/6                            581 ns        576 ns    1182552 6
Evaluate_SourceFunction/8                            582 ns        578 ns    1103248 8
Evaluate_SourceFunction/10                           585 ns        582 ns    1154792 10
Evaluate_SourceFunction/12                           597 ns        593 ns    1171254 12
Parse_Filter                                        4766 ns       4728 ns     149328
Parse_EvaluateFilter                                 222 ns        220 ns    3169271
TileMaskGeneration                                  1758 ns       1746 ns     396114
Parse_VectorTile                                 1956824 ns    1945068 ns        355
Util_dtoa                                           2090 ns       2073 ns     335116
Util_standardDtoa                                   2666 ns       2653 ns     257866
Util_dtoaLimits                                      311 ns        309 ns    2250334
Util_standardDtoaLimits                            36061 ns      35840 ns      19165
TileCountBounds                                       94 ns         94 ns    7198462
TileCountPolygon                                   23070 ns      22927 ns      30149
TileCoverPitchedViewport                            3407 ns       3384 ns     205752
TileCoverBounds                                     1281 ns       1273 ns     534572
TileCoverPolygon                                    9315 ns       9248 ns      73563
```